### PR TITLE
Manage length of custom field values

### PR DESCRIPTION
**Pull request by dadarya0**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | No
| New feature/enhancement? (use the a.x branch)      | Yes
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | No
| Automated tests included?              | Yes 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 

#### Description:
We're pretty consistently hitting custom field limits with our larger customers. It's because of a MySQL hard limit that we can't ultimately avoid. However, we can help mitigate it by adding a couple changes to Mautic. Currently we create all "text" fields with a 255 characters length and attempt to add an index to each.

We should implement the feature to add the ability to define the maximum number of characters for a "text" custom field. That way the customer can decide if the field only has values with a max of 10 characters which would then give 245 characters to another field.

We provide a "indexable" flag to the "text" custom field. If set to yes, they must define the limit as less than 255 characters and we can also add some validation to determine if we will hit MySQL limits for indexes (only 64 are allowed). If they have a field that is more than 255 characters and/or not indexable, then we can create the field as a MySQL text field instead of a varchar which is not supposed to go against the hard row length limit. We should also create textarea custom fields as MySQL text fields so they are not counted toward the length.

Decide what should be "Indexable" translated to
Show index flag on custom field list in last column as icon
Filter by indexed flag
Add indexable flag to text column field
use it when creating/editing custom field
Show in popup what "indexable means with available indexes count
Validate count of indexes when creating/editing custom fields
limit character length less than 255 via additional input if indexable
use it when updating data
Sync saved indexed w existing indexes
Change multi select to long text (no migration)
Don't allow editing length (but allow editing indexes)
Add migration to set text fields to 255 length (BC)

#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Run migrations, new columns is_index and char_length_limit in lead_fields table are created and char_length_limit is set to 255 for current text fields, is_index is set to true for current fields if the field has already index
3. When creating text fields (text, select, phone, url, email), you can use indexable flag to decide whether the index should be created and you can set field length, which will be used for column creation in the db. Default value can't be longer than character limit. Indexable flag has a tooltip with information about remaining available indexes.
4. You can change indexable flag for current fields, index will be updated in the db
5. While submitting contact, text fields are validated for a character limit
6. Filters are fixed for custom fields, you can filter also indexable or unique fields

**Add/Update custom field with both value true and false for create_custom_field_in_background flag**

### CODE REVIEW

File: app/bundles/CoreBundle/Doctrine/Helper/IndexSchemaHelper.php
Comment (escopecz) [inline app/bundles/CoreBundle/Doctrine/Helper/IndexSchemaHelper.php]:
```suggestion
    /**
     * @throws SchemaException
     */
    public function hasUniqueIdentifierIndex(LeadField $leadField): bool
```

File: app/bundles/LeadBundle/Field/Command/UpdateCustomFieldCommand.php
Comment (escopecz) [inline app/bundles/LeadBundle/Field/Command/UpdateCustomFieldCommand.php:2]:
```suggestion

declare(strict_types=1);

```

File: app/bundles/LeadBundle/Field/Command/UpdateCustomFieldCommand.php
Comment (escopecz) [inline app/bundles/LeadBundle/Field/Command/UpdateCustomFieldCommand.php]:
```suggestion
    protected function execute(InputInterface $input, OutputInterface $output): int
```

File: app/bundles/LeadBundle/Field/Event/UpdateColumnBackgroundEvent.php
Comment (escopecz) [inline app/bundles/LeadBundle/Field/Event/UpdateColumnBackgroundEvent.php]:
```suggestion
    public function getLeadField(): LeadField
```

File: app/bundles/LeadBundle/Field/Event/UpdateColumnBackgroundEvent.php
Comment (escopecz) [inline app/bundles/LeadBundle/Field/Event/UpdateColumnBackgroundEvent.php:2]:
```suggestion

declare(strict_types=1);

```

File: app/bundles/LeadBundle/Field/Event/UpdateColumnBackgroundEvent.php
Comment (escopecz) [inline app/bundles/LeadBundle/Field/Event/UpdateColumnBackgroundEvent.php]:
```suggestion
final class UpdateColumnBackgroundEvent extends Event
```

File: app/bundles/LeadBundle/Field/Event/UpdateColumnEvent.php
Comment (escopecz) [inline app/bundles/LeadBundle/Field/Event/UpdateColumnEvent.php:2]:
```suggestion

declare(strict_types=1);

```

File: app/bundles/LeadBundle/Field/Event/UpdateColumnEvent.php
Comment (escopecz) [inline app/bundles/LeadBundle/Field/Event/UpdateColumnEvent.php]:
```suggestion
final class UpdateColumnEvent extends Event
```

File: app/bundles/LeadBundle/Field/Exception/AbortColumnUpdateException.php
Comment (escopecz) [inline app/bundles/LeadBundle/Field/Exception/AbortColumnUpdateException.php:2]:
```suggestion

declare(strict_types=1);

```

File: app/bundles/LeadBundle/Field/Helper/IndexHelper.php
Comment (escopecz) [inline app/bundles/LeadBundle/Field/Helper/IndexHelper.php:2]:
```suggestion

declare(strict_types=1);

```

File: app/bundles/LeadBundle/Field/Helper/IndexHelper.php
Comment (escopecz) [inline app/bundles/LeadBundle/Field/Helper/IndexHelper.php]:
```suggestion
        $this->indexedColumns = array_map(
            fn ($index) => $index['Column_name'],
            $indexes
        );
```

File: app/bundles/LeadBundle/Form/Type/FieldType.php
Comment (escopecz) [inline app/bundles/LeadBundle/Form/Type/FieldType.php]:
```suggestion
    public static function validateDefaultValue(string $value, ExecutionContextInterface $context): void
```

File: app/bundles/LeadBundle/Tests/Field/Helper/IndexHelperTest.php
Comment (escopecz) [inline app/bundles/LeadBundle/Tests/Field/Helper/IndexHelperTest.php:2]:
```suggestion

declare(strict_types=1);

```

File: app/bundles/LeadBundle/Tests/Field/Dispatcher/FieldColumnDispatcherTest.php
Comment (escopecz) [inline app/bundles/LeadBundle/Tests/Field/Dispatcher/FieldColumnDispatcherTest.php]:
```suggestion
                $this->callback(fn (AddColumnBackgroundEvent $event) => $event instanceof UpdateColumnEvent),
```

**Review by escopecz — CHANGES_REQUESTED** [review]
Looks good. I suggested a bunch of improvements we can do now with PHP 8. You can accept those that you agree with and merge them all in one commit within the GH UI.

**Review by escopecz — CHANGES_REQUESTED** [review]
I tried to make the phone field indexable and this happened:
```
500 Internal Server Error - There is no column with name "email" on table "companies".
/html/vendor/doctrine/dbal/src/Schema/Exception/ColumnDoesNotExist.php:16 at
/html/vendor/doctrine/dbal/src/Schema/SchemaException.php:111 at Doctrine\DBAL\Schema\Exception\ColumnDoesNotExist :: new ( 'email', 'companies' )
/html/vendor/doctrine/dbal/src/Schema/Table.php:785 at Doctrine\DBAL\Schema\SchemaException :: columnDoesNotExist ( 'email', 'companies' )
/html/app/bundles/CoreBundle/Doctrine/Helper/IndexSchemaHelper.php:211 at Doctrine\DBAL\Schema\Table -> getColumn ( 'email' )
/html/app/bundles/CoreBundle/Doctrine/Helper/IndexSchemaHelper.php:90 at Mautic\CoreBundle\Doctrine\Helper\IndexSchemaHelper -> getTextColumns ( array( 'email' ) )
/html/app/bundles/LeadBundle/Field/CustomFieldIndex.php:154 at Mautic\CoreBundle\Doctrine\Helper\IndexSchemaHelper -> addIndex ( array( 'email' ), 'unique_identifier_search' )
/html/app/bundles/LeadBundle/Field/CustomFieldIndex.php:39 at Mautic\LeadBundle\Field\CustomFieldIndex -> updateUniqueIdentifierIndex ( object(LeadField) )
/html/app/bundles/LeadBundle/Field/CustomFieldColumn.php:135 at Mautic\LeadBundle\Field\CustomFieldIndex -> addIndexOnColumn ( object(LeadField) )
/html/app/bundles/LeadBundle/Field/CustomFieldColumn.php:51 at Mautic\LeadBundle\Field\CustomFieldColumn -> processUpdateLeadColumn ( object(LeadField) )
/html/app/bundles/LeadBundle/Model/FieldModel.php:559 at Mautic\LeadBundle\Field\CustomFieldColumn -> createLeadColumn ( object(LeadField) )
/html/app/bundles/LeadBundle/Controller/FieldController.php:297 at Mautic\LeadBundle\Model\FieldModel -> saveEntity ( object(LeadField), true )
/html/vendor/symfony/http-kernel/HttpKernel.php:163 at Mautic\LeadBundle\Controller\FieldController -> editAction ( object(Request), '32', false )
/html/vendor/symfony/http-kernel/HttpKernel.php:75 at Symfony\Component\HttpKernel\HttpKernel -> handleRaw ( object(Request), 2 )
/html/vendor/symfony/framework-bundle/Controller/AbstractController.php:156 at Symfony\Component\HttpKernel\HttpKernel -> handle ( object(Request), 2 )
/html/app/bundles/CoreBundle/Controller/CommonController.php:411 at Symfony\Bundle\FrameworkBundle\Controller\AbstractController -> forward ( 'Mautic\LeadBundle\Controller\FieldController::editAction', array( 'objectId' => '32', 'objectModel' => '', '_stopwatch_token' => '282b7f', '_route' => 'mautic_contactfield_action', '_controller' => 'Mautic\LeadBundle\Controller\FieldController::editAction', 'objectAction' => 'edit', '_route_params' => array( 'objectId' => '32', 'objectAction' => 'edit' ), '_firewall_context' => 'security.firewall.map.context.main', '_security_firewall_run' => '_security_mautic' ), array( 'mauticUserLastActive' => '97', 'mauticLastNotificationId' => '' ) )
/html/vendor/symfony/http-kernel/HttpKernel.php:163 at Mautic\CoreBundle\Controller\CommonController -> executeAction ( object(Request), 'edit', '32', 0, '' )
/html/vendor/symfony/http-kernel/HttpKernel.php:75 at Symfony\Component\HttpKernel\HttpKernel -> handleRaw ( object(Request), 1 )
/html/vendor/symfony/http-kernel/Kernel.php:202 at Symfony\Component\HttpKernel\HttpKernel -> handle ( object(Request), 1, true )
/html/app/AppKernel.php:109 at Symfony\Component\HttpKernel\Kernel -> handle ( object(Request), 1, true )
/html/app/middlewares/CORSMiddleware.php:76 at AppKernel -> handle ( object(Request), 1, true )
/html/app/middlewares/HSTSMiddleware.php:39 at Mautic\Middleware\CORSMiddleware -> handle ( object(Request), 1, true )
/html/app/middlewares/CatchExceptionMiddleware.php:28 at Mautic\Middleware\HSTSMiddleware -> handle ( object(Request), 1, true )
/html/app/middlewares/Dev/IpRestrictMiddleware.php:52 at Mautic\Middleware\CatchExceptionMiddleware -> handle ( object(Request), 1, true )
/html/app/middlewares/VersionCheckMiddleware.php:58 at Mautic\Middleware\Dev\IpRestrictMiddleware -> handle ( object(Request), 1, true )
/html/app/middlewares/TrustMiddleware.php:42 at Mautic\Middleware\VersionCheckMiddleware -> handle ( object(Request), 1, true )
/html/vendor/stack/builder/src/Stack/StackedHttpKernel.php:23 at Mautic\Middleware\TrustMiddleware -> handle ( object(Request), 1, true )
/html/index.php:19 at Stack\StackedHttpKernel -> handle ( object(Request) )
```
After that the phone field is set as indexable in the UI but I don't see the index in the leads table.

It worked well when creating a new field with specific length and index.

**Comment by dadarya0** [comment]
> I tried to make the phone field indexable and this happened:
> 
> ```
> 500 Internal Server Error - There is no column with name "email" on table "companies".
> /html/vendor/doctrine/dbal/src/Schema/Exception/ColumnDoesNotExist.php:16 at
> /html/vendor/doctrine/dbal/src/Schema/SchemaException.php:111 at Doctrine\DBAL\Schema\Exception\ColumnDoesNotExist :: new ( 'email', 'companies' )
> /html/vendor/doctrine/dbal/src/Schema/Table.php:785 at Doctrine\DBAL\Schema\SchemaException :: columnDoesNotExist ( 'email', 'companies' )
> /html/app/bundles/CoreBundle/Doctrine/Helper/IndexSchemaHelper.php:211 at Doctrine\DBAL\Schema\Table -> getColumn ( 'email' )
> /html/app/bundles/CoreBundle/Doctrine/Helper/IndexSchemaHelper.php:90 at Mautic\CoreBundle\Doctrine\Helper\IndexSchemaHelper -> getTextColumns ( array( 'email' ) )
> /html/app/bundles/LeadBundle/Field/CustomFieldIndex.php:154 at Mautic\CoreBundle\Doctrine\Helper\IndexSchemaHelper -> addIndex ( array( 'email' ), 'unique_identifier_search' )
> /html/app/bundles/LeadBundle/Field/CustomFieldIndex.php:39 at Mautic\LeadBundle\Field\CustomFieldIndex -> updateUniqueIdentifierIndex ( object(LeadField) )
> /html/app/bundles/LeadBundle/Field/CustomFieldColumn.php:135 at Mautic\LeadBundle\Field\CustomFieldIndex -> addIndexOnColumn ( object(LeadField) )
> /html/app/bundles/LeadBundle/Field/CustomFieldColumn.php:51 at Mautic\LeadBundle\Field\CustomFieldColumn -> processUpdateLeadColumn ( object(LeadField) )
> /html/app/bundles/LeadBundle/Model/FieldModel.php:559 at Mautic\LeadBundle\Field\CustomFieldColumn -> createLeadColumn ( object(LeadField) )
> /html/app/bundles/LeadBundle/Controller/FieldController.php:297 at Mautic\LeadBundle\Model\FieldModel -> saveEntity ( object(LeadField), true )
> /html/vendor/symfony/http-kernel/HttpKernel.php:163 at Mautic\LeadBundle\Controller\FieldController -> editAction ( object(Request), '32', false )
> /html/vendor/symfony/http-kernel/HttpKernel.php:75 at Symfony\Component\HttpKernel\HttpKernel -> handleRaw ( object(Request), 2 )
> /html/vendor/symfony/framework-bundle/Controller/AbstractController.php:156 at Symfony\Component\HttpKernel\HttpKernel -> handle ( object(Request), 2 )
> /html/app/bundles/CoreBundle/Controller/CommonController.php:411 at Symfony\Bundle\FrameworkBundle\Controller\AbstractController -> forward ( 'Mautic\LeadBundle\Controller\FieldController::editAction', array( 'objectId' => '32', 'objectModel' => '', '_stopwatch_token' => '282b7f', '_route' => 'mautic_contactfield_action', '_controller' => 'Mautic\LeadBundle\Controller\FieldController::editAction', 'objectAction' => 'edit', '_route_params' => array( 'objectId' => '32', 'objectAction' => 'edit' ), '_firewall_context' => 'security.firewall.map.context.main', '_security_firewall_run' => '_security_mautic' ), array( 'mauticUserLastActive' => '97', 'mauticLastNotificationId' => '' ) )
> /html/vendor/symfony/http-kernel/HttpKernel.php:163 at Mautic\CoreBundle\Controller\CommonController -> executeAction ( object(Request), 'edit', '32', 0, '' )
> /html/vendor/symfony/http-kernel/HttpKernel.php:75 at Symfony\Component\HttpKernel\HttpKernel -> handleRaw ( object(Request), 1 )
> /html/vendor/symfony/http-kernel/Kernel.php:202 at Symfony\Component\HttpKernel\HttpKernel -> handle ( object(Request), 1, true )
> /html/app/AppKernel.php:109 at Symfony\Component\HttpKernel\Kernel -> handle ( object(Request), 1, true )
> /html/app/middlewares/CORSMiddleware.php:76 at AppKernel -> handle ( object(Request), 1, true )
> /html/app/middlewares/HSTSMiddleware.php:39 at Mautic\Middleware\CORSMiddleware -> handle ( object(Request), 1, true )
> /html/app/middlewares/CatchExceptionMiddleware.php:28 at Mautic\Middleware\HSTSMiddleware -> handle ( object(Request), 1, true )
> /html/app/middlewares/Dev/IpRestrictMiddleware.php:52 at Mautic\Middleware\CatchExceptionMiddleware -> handle ( object(Request), 1, true )
> /html/app/middlewares/VersionCheckMiddleware.php:58 at Mautic\Middleware\Dev\IpRestrictMiddleware -> handle ( object(Request), 1, true )
> /html/app/middlewares/TrustMiddleware.php:42 at Mautic\Middleware\VersionCheckMiddleware -> handle ( object(Request), 1, true )
> /html/vendor/stack/builder/src/Stack/StackedHttpKernel.php:23 at Mautic\Middleware\TrustMiddleware -> handle ( object(Request), 1, true )
> /html/index.php:19 at Stack\StackedHttpKernel -> handle ( object(Request) )
> ```
> 
> After that the phone field is set as indexable in the UI but I don't see the index in the leads table.
> 
> It worked well when creating a new field with specific length and index.

Issue fixed by https://github.com/acquia/mc-cs/pull/767 and https://github.com/acquia/mc-cs/pull/1101

**Comment by escopecz** [comment]
@dadarya0 I still cannot create an index on the phone field. It's not throwing the error anymore. I tried also to run
```
bin/console mautic:custom-field:update-column -i 32 -u 1

Column was updated
```
but it didn't help. Is it working for you?

**Comment by dadarya0** [comment]
> bin/console mautic:custom-field:update-column

yes, it sis working for me, follow below steps :-
1. clear your cache and regenerate assets as 1 JS file is changed
2. Run migration
3. set 'create_custom_field_in_background' => false and test everything should work instantly
4. set 'create_custom_field_in_background' => true and test everything should work only after command execution.

**Comment by escopecz** [comment]
It works well on the custom text field I've created. I can add and remove index as I please. But on existing custom fields it doesn't create the index. For example on phone or city fields. Can you try that?

**Comment by kuzmany** [comment]
Almost everything works perfectly. It's a nice feature. However, there is one thing...
- same like John, core fields are not updated
- `mautic:custom-field:update-column`  You need the exact ID to make a call. It would be beneficial to periodically update the fields without parameters

**Review by escopecz — APPROVED** [review]
🤦 I was changing the index on the company field but I was looking into the leads table. It works!

@kuzmany let's add this change as a new PR. By default you don't need any command as it works during the field save. I'd like to improve this later with Symfony Messenger as we should create a job when a field is changed that will make the change in the background. Then this command is all we need.

**Comment by escopecz** [comment]
Oops, I didn't notice that this doesn't have a second review. @rohitpavaskar please test this closed PR post-merge so we have the quality check completed.

**Comment by kuzmany** [comment]
Another note: 
We have enabled the background option. 
After the update, you will receive an alert stating: "Your custom field is being created. We will notify you when it is complete."

**Review by rohitpavaskar — COMMENTED** [review]
Working as expected. Thanks!

**Pull request merged**
